docs/key-issues/multiple_architectures.md

```diff
@@ -0,0 +1,266 @@
+library name.
+
+The final binary artefact produced for Android distrobution uses this same
+directory convention. A "binary" on Android is an APK (Android Application
+Package) bundle; this is effectively a ZIP file with known metadata and
+structure; internally, there are subfolders for each supported CPU architecture.
+This APK is bundled into AAB (Android Application Bundle) format for upload to
+an app store; at time of installation, a CPU-specific APK is generated and
+provided to the end-user for installation.
+
+### Fat binaries
+
+Apple has taken the approach of "fat" binaries. A fat binary is a single
+executable or library artefact that contains code for multiple CPU
+architectures.
+
+Fat binaries can be compiled in two ways:
+
+1. **Single pass** Apple has modified their compiler tooling with flags that
+   allow the user to specify a single compilation command, and instruct the
+   compiler to generate multiple output architectures in the output binary
+2. **Multiple pass** After compiling a binary for each platform, Apple provides
+   a call named `lipo` to combine multiple single-architecture binaries into a
+   single fat binary that contains all platforms.
+
+At runtime, the operating system loads the binary slice for the current CPU
+architecture, and the linker loads the appropriate slice from the fat binary of
+any dynamic libraries.
+
+On macOS ARM hardware, Apple also provides Rosetta as a support mechanism; if a
+user tries to run an binary that doesn't contain an ARM64 slice, but *does*
+contain an x86-64 slice, the x86-64 slice will be converted at runtime into an
+ARM64 binary. Complications can occur when only *some* of the binary is being
+converted (e.g., if the binary being executed is fat, but a dynamic library
+isn't).
+
+To support the transition to Apple Silicon/M1 (ARM64), Python has introduced a
+`universal2` architecture target. This is effectively a "fat wheel" format; the
+`.dylib` files contained in the wheel are fat binaries containing both x86-64
+and ARM64 slices.
+
+iOS has an additional complication of requiring support for mutiple *ABIs* in
+addition to multiple CPU architectures. The ABI for the iOS simulator and
+physical iOS devices are different; however, ARM64 is a supported CPU
+architecture for both. As a result, it is not possible to produce a single fat
+library that supports both the iOS simulator and iOS devices. Apple provides an
+additional structure - the `XCFramework` - as a wrapper format for packaging
+libraries that need to span multiple ABIs. When developing an application for
+iOS, a developer will need to install binaries for both the simulator and
+physical devices.
+
+## Problems
+
+At present, the Python ecosystem almost exclusively uses the "multiple binary"
+solution. This serves the needs of Windows and Linux well, as it matches the
+way end-users interact with binaries.
+
+The `universal2` "fat wheel" solution also works well for macOS. The definition
```

Comment:
I have to rewrite this - it is counter to what's more or less the consensus opinion in scientific/numerical projects I think, which is that `universal2` is not a good idea, and certainly inferior to simply distributing thin `x86_64` and `arm64` wheels.

Reply:
I'm intrigued about this consensus. I've definitely noticed the trend to *not* package fat binaries for numpy et al, but I haven't been clear on *why* it's been necessary. If you're able to build an x86_64 wheel and an M1 wheel, I'm not aware of any technical reason that you couldn't produce a universal2 wheel - even if it's done as a "make 2 wheels and merge".

Reply:
Why does it have to (or even should be expected to) be a technical reason? There are many things that are technologically possible, that doesn't always make them a good idea.

Specifically, though, the consensus is that they are a bad user experience. For projects which are often already very large, it's seen as burdensome to make users download lots of data they don't want or need. If you have two thin wheels, it's also extra work you didn't ask for, to fuse them into universal2. Extra work which seemingly serves no purpose other than to waste the network bandwidth and disk storage space of users.

This isn't great even for small projects. SciPy is "only" 30mb. For projects that are hundreds of mb, the user experience would become prohibitively bad.

After paying all this inconvenience and wasted power/resources, what benefit is seen at the end? Do the wheels run better on those systems?

Reply:
Do they run better? For me, that's an unequivocal "yes, they do", because I can download a single macOS wheel with a single `pip install` command, and my macOS app will run on all Macs.

If the python/pip ecosystem had better support for multi-architecture deployment, then sure - single architecture wheels could be a better solution, requiring less download volume. However, absent that tooling, universal2 wheels provide a *much* better user experience for my users.

Reply:
> I've definitely noticed the trend to not package fat binaries for numpy et al, but I haven't been clear on why it's been necessary.

There's multiple reasons:
1. The tradeoff between download size and disk space usage vs. the upside for say a .dmg installer is (a) bad (we're talking 100s of MBs per Python env, and users/devs are likely to have quite a few envs, so this is GBs wasted - and macbook disk space isn't cheap ...), and (b) really does not exist for wheels or libraries to begin with - it's only applicable to end users downloading a .dmg and then not having to choose (which means, it applies mostly to non-technical end users only)
2. It imposes an extra burden on CI and PyPI total size usage for each project
3. When you anyway provide thin wheels, you cannot even install a universal2 wheel with `pip` from PyPI at all. Why upload artifacts you cannot install?

The reasoning given for universal2 wheels is extremely thin. There's no argument there that wouldn't equally apply to, say, 32-bit and 64-bit Windows wheels. Users get confused there too, more than on macOS even. But we don't stuff 32/64-bit builds into a single Windows wheel either. Nor for any other platform. The whole `universal2` thing is just bad and should probably never have been permitted on PyPI. The _only_ folks who _maybe_ need it are people building installers directly for nontechnical end users. And those can use `delocate-fuse` easily when building those installers. 

> If the python/pip ecosystem had better support for multi-architecture deployment,

Clearly you're having a problem here, but it seems like that's specific to iOS. For macOS laptop/desktop there should never be a need for multi-architecture deployment based on wheels.

Reply:
> Do they run better? For me, that's an unequivocal "yes, they do", because I can download a single macOS wheel with a single pip install command, and my macOS app will run on all Macs.

But they don't run better, even for you -- they run exactly the same.

What's better isn't running the wheels, only using developer tooling to craft those wheels into apps. End users don't transfer wheels between Macs, they only transfer apps.


> If the python/pip ecosystem had better support for multi-architecture deployment, then sure - single architecture wheels could be a better solution, requiring less download volume. However, absent that tooling, universal2 wheels provide a much better user experience for my users.

Given a choice between CPython developing tooling for multi-architecture deployment, or multi-architecture compilation, it baffles me that the powers that be chose the latter.

Reply:
This is "probably okay" if the goal is just to bake the binaries into an app, but not provide that app with the ability to compile user code. Of course the python ecosystem doesn't have a clear distinction between runtime and development code...

Reply:
@isuruf isn't that just a bug that was present because of `universal2` existing at all, and it's now fixed and should work just fine for fused thin wheels? See https://github.com/numpy/numpy/blob/main/numpy/core/include/numpy/numpyconfig.h#L6-L55 - the hardcoding is always there.

Second comment I have is that if one would only fuse the final artifacts that go into an app, something like this would never be a problem.

Reply:
Yes, it was fixed in `numpyconfig.h`. (`_numpyconfig.h` with the `_` prefix are different for the two architectures and the lines you mentioned fixes the issue with the divergence in `_numpyconfig.h`)

> Second comment I have is that if one would only fuse the final artifacts that go into an app, something like this would never be a problem.

In this particular case, yes it is not an issue because numpy's runtime behavior does not depend on the header file. However a package at runtime might depend on a generated header or a generated configuration file and it might be an issue. I haven't seen any package like that at the moment. Just an issue to keep in mind.

Reply:
> > It's probably a cultural thing
> Part of the issue is that Apple itself has been pushing for this kind of thing. Perhaps there's even a business perspective there that's anti-aligned with end user interest (Apple probably doesn't care about disk space usage, maybe they even sell beefier machines).

I strongly suspect that this is is indeed a cultural issue - but not the one you think it is.

Most users of computers *do not not know or care* what CPU their computer has. It is a utility device for getting a job done. To the extent that they are aware of CPU architecture during a laptop purchase, it will be a bullet point on the sales chart next to "rose gold unibody aluminium frame". 

In this environment, Apple has made the decision to trade off disk space (which is relatively cheap, plentiful, user-extensible, and recoverable) for user convenience. The user doesn't need to know or care what CPU they have - well behaved Apple apps run on *every* Apple machine.

Compare this with the Windows on ARM experience, where just this weekend, one of the attendees in the tutorial I was running was having difficulties, and it wasn't until I asked "wait - is that an ARM surface" that it occurred to him that this could be the source of the problems he was having. This is a class of user confusion that Mac users *do not have*. This was at a conference for Open Source enthusiasts, so it's hard to push the argument that this was a "non technical  user" problem.

Universal2 isn't the *only* solution to this problem. Having single architecture wheels *and* core packaging infrastructure that allows for installing multiple platform wheels for the same install would also be a viable solution to this problem. However, at present, the former exists, and the latter doesn't.

Reply:
> One thing to keep in mind is that merging two thin wheels might not make a universal2 with the desired functionality. For eg: numpy generated header files with the size of `long double` hardcoded and running `delocate-fuse` would pick the header from `x86_64` wheel which was wrong for the `arm64` build. (This was fixed by generating a header that works for both architectures.)

On the same theme - both wheels will contain a full copy of the Python code for the package. When you merge 2 binary wheels, which copy of the code do you keep? If there's any sort of conditional module inclusion in the packaging process (e.g., only including an `arm64.py` or `x86_64.py` depending on the target platform), or any sort of build-time (re)writing of modules (e.g., renaming `arm64.py` to `engine.py` as part of the build, those changes *also* need to be merged. 

I'm sure it's possible to come up with policy rules that allow for this sort of thing, but it's something that needs to be explicit.

Reply:
> When you merge 2 binary wheels, which copy of the code do you keep?

I would say: validate that the files are the same, and if they're not the same, error out. This problem exists whether universal2 or thin wheels are being built, though. Maybe it's a bit *easier* to catch with fusing thin wheels, however, since you'd sidestep the chance the build system will produce fat binaries but do pure-python generation based on whatever the current running platform seems to be.

Reply:
This merging strategy is a very good point indeed. It would have caught the bug that @isuruf pointed out. I think there's a bit more to it than "error out"; that'd be a good default but doesn't always work. E.g., numpy/scipy have a generated `__config__.py` containing build-time paths. The merge strategy there should be "random" not "error". If that could be specified per file, it'd also be good for the reproducible builds effort.

> Most users of computers _do not not know or care_ what CPU their computer has.

I agree. That doesn't translate to "users of wheels" though. 

> In this environment, Apple has made the decision to trade off disk space (which is relatively cheap, plentiful, user-extensible, and recoverable) for user convenience.

I'd say that that's a value judgement. It hasn't applied for me in the past - when I was a student and recent graduate without much money, I bought the minimum spec'd machine (128 GB disk, 4/8 GB RAM IIRC) and usually had to manage disk space very carefully. I still hear these complaints from users that buy a MacBook Air. A few hundred dollars more for a larger disk is a lot of money for many demographics. Same for bandwidth - we no longer care too much in our privileged positions, but in most countries in this world, unlimited plans are not common or not even available. So my value judgement here is: `universal2` wheels are quite dismissive of those concerns, and against the inclusiveness the Python community aims for.

> Having single architecture wheels and core packaging infrastructure that allows for installing multiple platform wheels for the same install would also be a viable solution to this problem. However, at present, the former exists, and the latter doesn't.

Both exist and both are incomplete I'd say (because you can't even install `universal2` wheels with `pip` when all thin/fat flavors are uploaded to PyPI - it seems really no one uses or cares much about `universal2` beyond a handful of people). Right now the `delocate-fuse` approach is _more_ incomplete, so let's add as a solution direction to address that?

Reply:
Oh. Speaking of which -- Apple made the curious decision to make storage space not, in fact, extensible. ;)

The hard drive is soldered in, you can't replace it with a larger capacity after buying the device. So much for user convenience.

If we're discussing why Apple makes decisions, I would suspect the real reason Apple made the universal2 decision is to get people to buy the comparatively expensive luxury devices with high-specced storage while simultaneously dodging the blame by making it seem like something you just do because of software distributors.

Reply:
@freakboy3742 I pushed a commit which adds a foldout with title _"what's the deal with `universal2`?"_. I provided arguments against; if you could add arguments in favor, that'd be great. Once we have that, we can update the current state and problem sections.